xla/service/gpu/gpu_all_gather_optimizer.cc

```diff
@@ -0,0 +1,194 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/gpu_all_gather_optimizer.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_casting_utils.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instruction.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instructions.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_module.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_opcode.h"
+#include "tensorflow/compiler/xla/service/collective_ops_utils.h"
+#include "tensorflow/compiler/xla/service/hlo_query.h"
+#include "tensorflow/compiler/xla/service/reduce_scatter_utils.h"
+
+namespace xla {
+namespace gpu {
+
+// Structure to keep track of the subgraph to optimize
+// "Left" and "Right" are conceptual. The initial
+// reduce scatter is always considered "left" and "right"
+// points to the other side of subgraph ending in a binary op
+
+struct AllGatherReduceScatterSpec {
+  HloInstruction* all_gather_left;
+  HloInstruction* all_gather_right;
+  HloInstruction* right_source;
+  HloInstruction* binary_op;
+};
+
+std::optional<AllGatherReduceScatterSpec> MatchReduceScatter(
+    const HloReduceScatterInstruction* rs, int64_t min_rank = 1) {
+  if (!rs->shape().IsArray() || rs->constrain_layout() ||
+      (rs->IsCrossModuleAllReduce() &&
+       !rs->GetModule()->config().use_spmd_partitioning())) {
+    VLOG(2) << "Unsupported reduce-scatter: " << rs->ToString();
+    return std::nullopt;
+  }
+  if (rs->shape().rank() - absl::c_count(rs->shape().dimensions(), 1) <
+      min_rank) {
+    VLOG(2) << " Should be at least rank-" << min_rank
+            << " excluding trivial dimensions " << rs->ToString();
+    return std::nullopt;
+  }
+  if (rs->replica_groups().size() > 1) {
+    const int64_t size = rs->replica_groups()[0].replica_ids_size();
+    absl::Span<const ReplicaGroup> rgs = rs->replica_groups();
+    const bool has_uniform_size = absl::c_all_of(
+        rgs.subspan(1, size - 1), [size](const ReplicaGroup& group) {
+          return group.replica_ids_size() == size;
+        });
+    if (!has_uniform_size) {
+      VLOG(2) << "Unsupported non-uniform replica group size "
+              << rs->ToString();
+      return std::nullopt;
+    }
+  }
+
+  HloInstruction* user = rs->users()[0];
+  if (user->opcode() != HloOpcode::kAllGather) {
+    // ideally this should cover cases with other ops in between
+    VLOG(2) << "Reduce-Scatter is not followed by all-gather " << user->ToString();
+    return std::nullopt;
+  }
+  if (user->user_count() != 1) {
+    // all-gather is subject to removal. Should not have more than 1 user
+    VLOG(2) << "all-gather user_count > 1 " << user->ToString();
+    return std::nullopt;
+  }
+  HloInstruction* allgatherLeft = user;
```

Comment:
Please use variable names like all_gather_left, not camel case.

Reply:
done.